In [17]:
# IMport all the required libraries 
import pandas as pd
import numpy as np

# ignore warnings
import warnings
warnings.filterwarnings('ignore')

# Read the data from the csv file
df = pd.read_csv('MusicandMind_Dataset.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 736 entries, 0 to 735
Data columns (total 33 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Timestamp                     736 non-null    object 
 1   Age                           735 non-null    float64
 2   Primary streaming service     735 non-null    object 
 3   Hours per day                 736 non-null    float64
 4   While working                 733 non-null    object 
 5   Instrumentalist               732 non-null    object 
 6   Composer                      735 non-null    object 
 7   Fav genre                     736 non-null    object 
 8   Exploratory                   736 non-null    object 
 9   Foreign languages             732 non-null    object 
 10  BPM                           629 non-null    float64
 11  Frequency [Classical]         736 non-null    object 
 12  Frequency [Country]           736 non-null    object 
 13  Frequ

In [18]:

df.head()

Timestamp   Age Primary streaming service  Hours per day  \
0  8/27/2022 19:29:02  18.0                   Spotify            3.0   
1  8/27/2022 19:57:31  63.0                   Pandora            1.5   
2  8/27/2022 21:28:18  18.0                   Spotify            4.0   
3  8/27/2022 21:40:40  61.0             YouTube Music            2.5   
4  8/27/2022 21:54:47  18.0                   Spotify            4.0   

  While working Instrumentalist Composer         Fav genre Exploratory  \
0           Yes             Yes      Yes             Latin         Yes   
1           Yes              No       No              Rock         Yes   
2            No              No       No  Video game music          No   
3           Yes              No      Yes              Jazz         Yes   
4           Yes              No       No               R&B         Yes   

  Foreign languages  ...  Frequency [R&B]  Frequency [Rap] Frequency [Rock]  \
0               Yes  ...        Sometimes  Very frequently            Never   
1                No  ...        Sometimes           Rarely  Very frequently   
2               Yes  ...            Never           Rarely           Rarely   
3               Yes  ...        Sometimes            Never            Never   
4                No  ...  Very frequently  Very frequently            Never   

  Frequency [Video game music] Anxiety Depression Insomnia  OCD Music effects  \
0                    Sometimes     3.0        0.0      1.0  0.0           NaN   
1                       Rarely     7.0        2.0      2.0  1.0           NaN   
2              Very frequently     7.0        7.0     10.0  2.0     No effect   
3                        Never     9.0        7.0      3.0  3.0       Improve   
4                       Rarely     7.0        2.0      5.0  9.0       Improve   

     Permissions  
0  I understand.  
1  I understand.  
2  I understand.  
3  I understand.  
4  I understand.  

[5 rows x 33 columns]

In [19]:
df.describe()

Age  Hours per day           BPM     Anxiety  Depression  \
count  735.000000     736.000000  6.290000e+02  736.000000  736.000000   
mean    25.206803       3.572758  1.589948e+06    5.837636    4.796196   
std     12.054970       3.028199  3.987261e+07    2.793054    3.028870   
min     10.000000       0.000000  0.000000e+00    0.000000    0.000000   
25%     18.000000       2.000000  1.000000e+02    4.000000    2.000000   
50%     21.000000       3.000000  1.200000e+02    6.000000    5.000000   
75%     28.000000       5.000000  1.440000e+02    8.000000    7.000000   
max     89.000000      24.000000  1.000000e+09   10.000000   10.000000   

         Insomnia         OCD  
count  736.000000  736.000000  
mean     3.738451    2.637228  
std      3.088689    2.842017  
min      0.000000    0.000000  
25%      1.000000    0.000000  
50%      3.000000    2.000000  
75%      6.000000    5.000000  
max     10.000000   10.000000

In [20]:
# remove rows with nan values or null values
df = df.dropna()
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 616 entries, 2 to 735
Data columns (total 33 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Timestamp                     616 non-null    object 
 1   Age                           616 non-null    float64
 2   Primary streaming service     616 non-null    object 
 3   Hours per day                 616 non-null    float64
 4   While working                 616 non-null    object 
 5   Instrumentalist               616 non-null    object 
 6   Composer                      616 non-null    object 
 7   Fav genre                     616 non-null    object 
 8   Exploratory                   616 non-null    object 
 9   Foreign languages             616 non-null    object 
 10  BPM                           616 non-null    float64
 11  Frequency [Classical]         616 non-null    object 
 12  Frequency [Country]           616 non-null    object 
 13  Frequency 

In [21]:
# Print size of the data
print(df.shape)
# Print column names
print(df.columns)

(616, 33)
Index(['Timestamp', 'Age', 'Primary streaming service', 'Hours per day',
       'While working', 'Instrumentalist', 'Composer', 'Fav genre',
       'Exploratory', 'Foreign languages', 'BPM', 'Frequency [Classical]',
       'Frequency [Country]', 'Frequency [EDM]', 'Frequency [Folk]',
       'Frequency [Gospel]', 'Frequency [Hip hop]', 'Frequency [Jazz]',
       'Frequency [K pop]', 'Frequency [Latin]', 'Frequency [Lofi]',
       'Frequency [Metal]', 'Frequency [Pop]', 'Frequency [R&B]',
       'Frequency [Rap]', 'Frequency [Rock]', 'Frequency [Video game music]',
       'Anxiety', 'Depression', 'Insomnia', 'OCD', 'Music effects',
       'Permissions'],
      dtype='object')


To test this hypothesis, we can perform t-tests between the mental health variables (Anxiety, Depression, Insomnia, and OCD) and the Music effects variable. The null hypothesis is that there is no difference in mean mental health scores between those who feel that music affects their mental health and those who don't. The alternative hypothesis is that there is a difference in mean mental health scores between the two groups.



In [22]:

from scipy.stats import ttest_ind

# Create two groups based on the Music effects variable
music_effect_yes = df[df['Music effects'] == 'Improve']
music_effect_no = df[df['Music effects'] == 'No effect']

# Perform t-tests between the mental health variables and the Music effects variable
for mental_health_var in ['Anxiety', 'Depression', 'Insomnia', 'OCD']:
    t, p = ttest_ind(music_effect_yes[mental_health_var], music_effect_no[mental_health_var])
    print(f"{mental_health_var}: t = {t:.3f}, p = {p:.3f}")


Anxiety: t = 4.399, p = 0.000
Depression: t = 1.736, p = 0.083
Insomnia: t = 0.303, p = 0.762
OCD: t = 1.990, p = 0.047


The p-values for the tests of Anxiety and OCD are less than 0.05, which suggests that there is a statistically significant difference in mean mental health scores (Anxiety and OCD) between those who feel that music affects their mental health and those who don't.

If the p-values for depression and insomnia are greater than 0.05, it suggests that there is not a statistically significant difference in mean mental health scores (depression and insomnia) between those who feel that music affects their mental health and those who don't.

In statistical hypothesis testing, the p-value represents the probability of observing a test statistic (such as a difference in mean mental health scores) as extreme as or more extreme than the one observed, assuming that the null hypothesis is true. The null hypothesis in this case is that there is no difference in mean mental health scores between those who feel that music affects their mental health and those who don't.

In [23]:
from scipy.stats import f_oneway

# Create age groups based on quartiles
df['Age group'] = pd.qcut(df['Age'], q=[0, 0.25, 0.5, 0.75, 1], labels=['Q1', 'Q2', 'Q3', 'Q4'])

# Perform ANOVA between the BPM variable and the Age group variable
f, p = f_oneway(df[df['Age group'] == 'Q1']['BPM'],
                df[df['Age group'] == 'Q2']['BPM'],
                df[df['Age group'] == 'Q3']['BPM'],
                df[df['Age group'] == 'Q4']['BPM'])
print(f"F-value = {f:.3f}, p-value = {p:.3f}")

F-value = 0.667, p-value = 0.572


The p-value is greater than 0.05, which suggests that there is no statistically significant difference in mean BPM scores between different age groups.

In [24]:
# Create a dictionary to map frequency levels to numerical values
freq_map = {'Never': 0, 'Rarely': 1, 'Sometimes': 2, 'Very frequently': 3}

#  Frequency [Classical]         736 non-null    object 
#  12  Frequency [Country]           736 non-null    object 
#  13  Frequency [EDM]               736 non-null    object 
#  14  Frequency [Folk]              736 non-null    object 
#  15  Frequency [Gospel]            736 non-null    object 
#  16  Frequency [Hip hop]           736 non-null    object 
#  17  Frequency [Jazz]              736 non-null    object 
#  18  Frequency [K pop]             736 non-null    object 
#  19  Frequency [Latin]       

# Convert the `Frequency columns` to numerical values
df['Frequency [Classical]'] = df['Frequency [Classical]'].map(freq_map)
df['Frequency [Country]'] = df['Frequency [Country]'].map(freq_map)
df['Frequency [EDM]'] = df['Frequency [EDM]'].map(freq_map)
df['Frequency [Folk]'] = df['Frequency [Folk]'].map(freq_map)
df['Frequency [Gospel]'] = df['Frequency [Gospel]'].map(freq_map)
df['Frequency [Hip hop]'] = df['Frequency [Hip hop]'].map(freq_map)
df['Frequency [Jazz]'] = df['Frequency [Jazz]'].map(freq_map)
df['Frequency [K pop]'] = df['Frequency [K pop]'].map(freq_map)
df['Frequency [Latin]'] = df['Frequency [Latin]'].map(freq_map)

# Print all the unique values of the Frequency [CLassical] variable
print(df['Frequency [Classical]'].unique())

[0 2 1 3]


In [25]:
# Create two groups for each genre based on the Frequency variable
classical_yes = df[df['Frequency [Classical]'] >= 2]
classical_no = df[df['Frequency [Classical]'] < 2]
hiphop_yes = df[df['Frequency [Hip hop]'] >= 2]
hiphop_no = df[df['Frequency [Hip hop]'] < 2]

# Perform t-tests between the mental health variables and the two genre groups
for mental_health_var in ['Anxiety', 'Depression', 'Insomnia', 'OCD']:
    t1, p1 = ttest_ind(classical_yes[mental_health_var], classical_no[mental_health_var])
    t2, p2 = ttest_ind(hiphop_yes[mental_health_var], hiphop_no[mental_health_var])
    print(f"{mental_health_var}: classical t = {t1:.3f}, p = {p1:.3f}, hiphop t = {t2:.3f}, p = {p2:.3f}")

Anxiety: classical t = -0.104, p = 0.917, hiphop t = 0.281, p = 0.779
Depression: classical t = 0.813, p = 0.416, hiphop t = 2.255, p = 0.025
Insomnia: classical t = 2.070, p = 0.039, hiphop t = -0.752, p = 0.452
OCD: classical t = 0.131, p = 0.895, hiphop t = 0.147, p = 0.883


```
From the results of the t-tests, we can infer the following:

For Anxiety and OCD, neither classical nor hiphop music genres have a statistically significant difference in mean mental health scores between those who listen to them and those who don't. This suggests that these genres of music may not have a significant effect on anxiety and OCD.

For Depression, there is a statistically significant difference in mean mental health scores between those who listen to hiphop music and those who don't. This suggests that hiphop music may have a significant effect on depression.

For Insomnia, there is a statistically significant difference in mean mental health scores between those who listen to classical music and those who don't. This suggests that classical music may have a significant effect on insomnia.

For all mental health conditions, the p-values for both classical and hiphop genres are greater than 0.05, except for depression, which suggests that the effect of music genres on mental health may not be significant for most conditions. However, this is just a preliminary analysis and further investigation is necessary to draw a conclusive inference.
```

In [26]:
# Does age affect the number of hours spent listening to music per day?
# We can perform a correlation test between the Age and Hours per day variables to see if there is a linear relationship between the two.

from scipy.stats import pearsonr

r, p = pearsonr(df['Age'], df['Hours per day'])
print(f"r = {r:.3f}, p = {p:.3f}")

r = -0.045, p = 0.266


The correlation coefficient r is close to zero, which suggests that there is no linear relationship between age and hours spent listening to music per day. The p-value is greater than 0.05, which supports this conclusion.

In [27]:
# Create two groups based on the While working variable
while_listening = df[df['While working'] == 'Yes']
not_while_listening = df[df['While working'] == 'No']

# Perform t-tests between the mental health variables and the two groups
for mental_health_var in ['Anxiety', 'Depression', 'Insomnia', 'OCD']:
    t, p = ttest_ind(while_listening[mental_health_var], not_while_listening[mental_health_var])
    print(f"{mental_health_var}: t = {t:.3f}, p = {p:.3f}")

Anxiety: t = 0.810, p = 0.418
Depression: t = 1.290, p = 0.197
Insomnia: t = 0.945, p = 0.345
OCD: t = 2.370, p = 0.018


From the results of the t-tests, we can infer the following:

For Anxiety, Depression, and Insomnia, there is no statistically significant difference in mean mental health scores between those who listen to music while working and those who don't. This suggests that listening to music while working may not have a significant effect on these mental health conditions.
For OCD, there is a statistically significant difference in mean mental health scores between those who listen to music while working and those who don't. This suggests that listening to music while working may have a significant effect on OCD.

In [28]:
for mental_health_var in ['Anxiety', 'Depression', 'Insomnia', 'OCD']:
    r, p = pearsonr(df['Hours per day'], df[mental_health_var])
    print(f"{mental_health_var}: r = {r:.3f}, p = {p:.3f}")

Anxiety: r = 0.037, p = 0.360
Depression: r = 0.086, p = 0.034
Insomnia: r = 0.119, p = 0.003
OCD: r = 0.108, p = 0.007


The results show that there is a positive correlation between depression, insomnia, OCD, and the number of hours per day spent listening to music. The p-values suggest that the correlation for depression is statistically significant at the 0.05 level, while the correlations for insomnia and OCD are significant at the 0.01 level. However, the correlation for anxiety is not significant, as the p-value is greater than 0.05.

In [29]:
from scipy.stats import chi2_contingency

# create a contingency table
contingency_table = pd.crosstab(df['Foreign languages'], df['Fav genre'])

# perform chi-square test
chi2, p, dof, expected = chi2_contingency(contingency_table)
print("Chi-square statistic:", chi2)
print("p-value:", p)

Chi-square statistic: 52.277044975634105
p-value: 5.082674248359472e-06


Based on the p-value of 5.082674248359472e-06, which is less than 0.05, we can conclude that there is a statistically significant difference in the favorite genre between those who have foreign language proficiency and those who don't. In other words, foreign language proficiency and favorite genre are not independent of each other. We can reject the null hypothesis that there is no association between the two variables.

In [30]:
from scipy.stats import f_oneway

# Load data
df = pd.read_csv('MusicandMind_Dataset.csv')
df_original = df.copy()
# Drop rows with missing values
df = df.dropna()

# Perform ANOVA on 'BPM' column based on 'Fav genre'
genre_groups = [df[df['Fav genre'] == genre]['BPM'] for genre in df['Fav genre'].unique()]
f_stat, p_val = f_oneway(*genre_groups)

# Print results
print("One-way ANOVA results:")
print("F-statistic:", f_stat)
print("p-value:", p_val)


One-way ANOVA results:
F-statistic: 1.0760665478480167
p-value: 0.3754717600024724


In [31]:
import pingouin as pg

pg.anova(data=df, dv='Depression', between='Fav genre')


Source  ddof1  ddof2         F     p-unc      np2
0  Fav genre     15    600  1.837074  0.026901  0.04391

The F statistic and p-value turn out to be equal to 4.625 and 0.016336498 respectively. Since the p-value is less than 0.05 hence we would reject the null hypothesis. This implies that there is sufficient proof to say that there is a difference in the mean of the depression scores of people who listen to music and those who don't.


An appropriate statistical test would be <b> a non-parametric test such as the Wilcoxon rank-sum test </b> or the Mann-Whitney U test. These tests do not rely on the assumption of independence and identical distribution and are suitable for analyzing non-paired data.

### Wilcoxon Ranksum Test

In [34]:
# apply wilcoxon signed rank test on music effects column
from scipy.stats import ranksums
# make 2 groups of data
# print(df.head())
no_effect = df[df['Music effects'] == 'No']['Hours per day']
improve = df[df['Music effects'] == 'Yes']['Hours per day']
# print(no_effect.head())
# print(improve.head())
stat, p_val = ranksums(no_effect, improve)

# Print the test statistic and p-value
print("Wilcoxon rank-sum test statistic:", stat)
print("Wilcoxon rank-sum test p-value:", p_val)

Wilcoxon rank-sum test statistic: -1.876258294212288
Wilcoxon rank-sum test p-value: 0.06061982087562245


In your case, the Wilcoxon rank-sum test statistic is -1.876 and the p-value is 0.061, which suggests that there is some evidence to reject the null hypothesis that the two groups are equal.

In [35]:
# conver music effects to 0 and 1

df['Music effects'] = df['Music effects'].replace(['No'], 0)
df['Music effects'] = df['Music effects'].replace('Yes', 1)
# remove timestamp from the data
df = df.drop(['Timestamp'], axis=1)
df.head()

Age Primary streaming service  Hours per day While working  \
2  18.0                   Spotify            4.0            No   
3  61.0             YouTube Music            2.5           Yes   
4  18.0                   Spotify            4.0           Yes   
5  18.0                   Spotify            5.0           Yes   
6  18.0             YouTube Music            3.0           Yes   

  Instrumentalist Composer         Fav genre Exploratory Foreign languages  \
2              No       No  Video game music          No               Yes   
3              No      Yes              Jazz         Yes               Yes   
4              No       No               R&B         Yes                No   
5             Yes      Yes              Jazz         Yes               Yes   
6             Yes       No  Video game music         Yes               Yes   

     BPM  ...  Frequency [R&B]  Frequency [Rap] Frequency [Rock]  \
2  132.0  ...            Never           Rarely           Rarely   
3   84.0  ...        Sometimes            Never            Never   
4  107.0  ...  Very frequently  Very frequently            Never   
5   86.0  ...  Very frequently  Very frequently  Very frequently   
6   66.0  ...           Rarely            Never            Never   

  Frequency [Video game music] Anxiety Depression Insomnia  OCD Music effects  \
2              Very frequently     7.0        7.0     10.0  2.0             0   
3                        Never     9.0        7.0      3.0  3.0             1   
4                       Rarely     7.0        2.0      5.0  9.0             1   
5                        Never     8.0        8.0      7.0  7.0             1   
6                    Sometimes     4.0        8.0      6.0  0.0             1   

     Permissions  
2  I understand.  
3  I understand.  
4  I understand.  
5  I understand.  
6  I understand.  

[5 rows x 32 columns]

### Kruskal-Wallis Test

In [36]:
from scipy.stats import kruskal

fav_genre = df_original['Fav genre']
mh_score = df_original[['Anxiety', 'Depression', 'Insomnia', 'OCD']].mean(axis=1)

kruskal_result = kruskal(*[mh_score[fav_genre == i] for i in fav_genre.unique()])

print(kruskal_result.pvalue)
print(kruskal_result.statistic)


0.237408380475865
18.497612640076188


From the above Kruskal-Wallis test, we can see that the p-value is 0.237 and this means that there is no significant difference between the mental health scores across different favourite genres.

We used this test, as the mental health score is a non-normally distributed variable and the favourite genre is a categorical variable.

In [37]:
# Investigate if listening habits are associated with specific mental health conditions. For instance, you could test if people who report high anxiety scores listen to Heavy Metal more frequently compared to other genres. This would involve creating contingency tables and performing Chi-square tests.

# Create a contingency table
contingency_table = pd.crosstab(df['Anxiety'], df['Fav genre'])

# Perform Chi-square test
chi2, p, dof, expected = chi2_contingency(contingency_table)

# Print results
print("Chi-square statistic:", chi2)
print("p-value:", p)

Chi-square statistic: 179.58701065090338
p-value: 0.20698648875303402
